File: cpp/3027.find-the-number-of-ways-to-place-people-ii.cpp
#include <vector>
using std::vector;

class Solution {
public:
  int numberOfPairs(vector<vector<int>> &points) {
    const int n = points.size();
    int count = 0;

    // sort x ascending, y descending
    std::sort(points.begin(), points.end(),
              [](const vector<int> &a, const vector<int> &b) -> bool {
                return a[0] < b[0] || (a[0] == b[0] && a[1] > b[1]);
              });

    int maxY, minY, negInf = std::numeric_limits<int>::min();

    for (int i = 0; i < n; i++) {
      maxY = points[i][1];
      minY = negInf;

      for (int j = i + 1; j < n; j++) {
        if (minY < points[j][1] && points[j][1] <= maxY) {
          count++;
          minY = points[j][1];
        }

        if (maxY == minY) {
          break;
        }
      }
    }

    return count;
  }
};

File: python/3027.find-the-number-of-ways-to-place-people-ii.py
class Solution:
    def numberOfPairs(self, points: list[list[int]]) -> int:
        points.sort(key=lambda x: (x[0], -x[1]))

        n = len(points)
        count = 0

        for i in range(n):
            maxY = points[i][1]
            minY = -float("inf")

            for j in range(i + 1, n):
                if minY < points[j][1] <= maxY:
                    count += 1
                    minY = points[j][1]

                if maxY == minY:
                    break

        return count

